rfcs/20191016-dlpack-support.md

```diff
@@ -0,0 +1,114 @@
+print(t2)
+dlpack = thdlpack.to_dlpack(t2)  # th tensor -> dlpack
+t3 = tfdlpack.from_dlpack(dlpack)  # dlpack -> tf tensor
+print(t3)
+```
+You will find that t1, t2 and t3 all have the same values, shape, and device contexts.
+Package dependency: tensorflow>=2.0
+
+Proposed code of converting a Tensorflow tensor to Torch tensor using DLPack natively:
+```python
+import numpy as np
+import tensorflow as tf
+import tensorflow.???.dlpack as tfdlpack
+import torch.utils.dlpack as thdlpack
+
+
+t1 = tf.constant([1, 2, 3], dtype=np.float32)
+dlpack = tfdlpack.to_dlpack(t1)  # tf tensor -> dlpack
+t2 = thdlpack.from_dlpack(dlpack)  # dlpack -> th tensor
+print(t2)
+dlpack = thdlpack.to_dlpack(t2)  # th tensor -> dlpack
+t3 = tfdlpack.from_dlpack(dlpack)  # dlpack -> tf tensor
+print(t3)
+```
+
+Package implementation details:
+The first design consideration is that we want to avoid any modification to the main Tensorflow library, so to get around the potential long delay of PR, code review, and release cycle of Tensorflow main package. Inspired by the solution from https://github.com/tobegit3hub/tftvm, we decide to implement the functionality as two custom tensor ops: to_dlpack and from_dlpack.
+
+Besides, we want this feature to be plugged into other projects quite easily. For example, any project that relies on this feature is able to run without compiling against Tensorflow's header files. Not only that an extra dependency usually means extra effort, but also that such maintenance is repetitive and should be handled by the feature developer (i.e., us) alone. To this end, we have an idea of releasing it as a python package. However, the question is how to invoke the two custom tensor ops in python? The challenge is that Tensorflow's custom op interface has a limited support of argument and return types, while to_dlpack and from_dlpack should have an argument/return type of DLPack object. We work around this by encoding the address of an DLPack object as an integer, so it can be accepted/returned by the custom op interface. Then, we decode it in python or C depending on whether we return it (to_dlpack) or consume it (from_dlpack).
+
+Finally, to achieve the maximal efficiency, we want the conversion happens without memory copy.
+
+For to_dlpack, the returned DLPack tensor shares the same memory address of the input Tensorflow tensor and holds a reference to it. Upon the destruction of the DLPack tensor, it will dereference the Tensorflow tensor, so it can be collected by Tensorflow's memory management. (inspired by PyTorch's DLPack implementation).
+For from_dlpack, it first creates an allocator object (subclass Tensorflow's allocator interface) that holds the reference to the DLPack tensor. The AllocateRaw function directly returns the memory it holds without creating any new buffer. Upon destruction, the DeallocateRaw function just calls the deletor of the DLPack tensor. (inspired by Tensorflow's immutable_constant_op).
```

Comment:
If you're in TensorFlow, you don't need to go through this indirection.  You could subclass `tensorflow::TensorBuffer` to implement the behavior you want.

Reply:
The main reason uses allocator is due to the alignment issue. However I found the alignment requirement is always 64 bytes later. Will this change in the future? 
Also I agree`TensorBuffer` is simpler and do you think it's fine to always use 64 bytes alignment requirement in dlpack operation?

Reply:
I don't see how alignment is related to using `TensorBuffer` or not.  IIUC in from_dlpack you roughly do this:

```
if (IsAligned(64, dlm_tensor->dl_tensor.data)) {
  // forward dlpack data without memcpy
} else {
  // allocate output buffer and memcpy
}
```

This seems correct to me and we'll continue doing this when we use a `TensorBuffer` subclass.

Reply:
You are right. I realized alignment is always 64 bytes after I finished my first version which used allocator.